src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexQueryPlan.java

```diff
@@ -74,36 +71,108 @@ private StorageAttachedIndexQueryPlan(ColumnFamilyStore cfs,
+        if (!selectedIndexes(rowFilter.root(), allIndexes, selectedIndexes))
+            return null;
+
+        // collect the features of the selected indexes
         IndexFeatureSet.Accumulator accumulator = new IndexFeatureSet.Accumulator();
+        for (StorageAttachedIndex index : selectedIndexes)
+            accumulator.accumulate(index.getIndexContext().indexFeatureSet());
+
+        return new StorageAttachedIndexQueryPlan(cfs,
+                                                 queryMetrics,
+                                                 rowFilter,
+                                                 ImmutableSet.copyOf(selectedIndexes),
+                                                 accumulator.complete());
+    }
 
-        for (RowFilter.Expression expression : rowFilter)
+    /**
+     * Collects the indexes that can be used with the specified filtering tree without doing a full index scan.
+     * </p>
+     * The selected indexes are those that can satisfy at least one of the expressions of the filter, and that
+     * aren't part of an OR operation that contains not indexed expressions, unless that OR operation is nested inside
+     * an AND operation that has at least one indexed operation.
+     * </p>
+     * For example, for {@code x AND y} we can use any index in {@code x}, {@code y}, or both.
+     * </p>
+     * For {@code x OR y} we can't use a single index on {@code x} or {@code y} because we would need to do a full index
+     * scan because of the unidexed expression. However, if both columns were indexed, we could use those two indexes.
+     * </p>
+     * For {@code (x OR y) AND z}, where {@code x} and {@code z} are indexed, we can use the index on {@code z}, even
+     * though we will ignore the index on {@code x}.
+     *
+     * @param element a row filter tree node
+     * @param allIndexes all the indexes in the index group
+     * @param selectedIndexes the set of indexes where we'll add those indexes can be used with the specified expression
+     * @return {@code true} if this has collected any indexes, {@code false} otherwise
+     */
+    private static boolean selectedIndexes(RowFilter.FilterElement element,
+                                           Set<StorageAttachedIndex> allIndexes,
+                                           Set<StorageAttachedIndex> selectedIndexes)
+    {
+        if (element.isDisjunction()) // OR, all restrictions should have an index
         {
-            // we ignore user-defined expressions here because we don't have a way to translate their #isSatifiedBy
-            // method, they will be included in the filter returned by QueryPlan#postIndexQueryFilter()
-            if (expression.isUserDefined())
-                continue;
-
-            acceptedExpressions.add(expression);
-            for (StorageAttachedIndex index : indexes)
+            Set<StorageAttachedIndex> orIndexes = new HashSet<>();
```

Comment:
minor performance suggestion: you could get rid of this allocation and later copying and just add directly to selectedIndexes collection.

Reply:
We only add the indexes for the `OR` components if all the components have an index. We must collect them somewhere until we know they all have a backing index.